strings.go

```diff
@@ -306,3 +315,57 @@ func keyStmt(key []byte) *ast.GenDecl {
 		}},
 	}
 }
+
+func strConstBlacklist(node ast.Node, info *types.Info, blacklist map[types.Object]struct{}) map[types.Object]struct{} {
```

Comment:
Unfortunately, this is too naive. You've just written specific code to handle two specific cases, but there will be more cases you haven't covered for sure. Here are some more examples:

```
var _ [mustBeConst]byte
_ = [5]string{mustBeConst: "foo"}
```

See https://golang.org/ref/spec for reference. Try searching for "constant representable".

Here is how I think we should implement this:

* Detect what expressions *must* be constants. These are array type lengths, value expressions in a `const` declaration, keys in an array composite literal, and so on.
* When such expressions are detected, walk their node sub-trees.
* If any identifier or literal is found in those node sub-trees, they must be blacklisted from being turned to variables.

I'm pretty positive that that strategy would be more generic and resilient than the one you have here. For example, the code you wrote to handle `const length = len(x)` is too naive, so it will easily fail with inputs like `const length = len(x)+len(y)`.

Reply:
Thanks for the feedback,  your approach seems reasonable will try to implement. :+1: